### PR TITLE
[Feature] Show more exercise information in the ExerciseGroupComponent

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
We want to display more information about the exercises in the exercise group view. This should include the diagram type, the repositories and buildplan URLs, similar to how it is displayed in the course management view

### Description
- [x] [Client] Extend the ExerciseGroupComponent HTML table to include more details about the exercises
- [x] [Client] Only show columns relevant to the existing exercises
- [x] [Client] Fixed a bug on text exercise creation where the assessment type is undefined if the corresponding checkbox hasn't been touched
- [x] [Server] Attach the template and solution participation to the programming exercises
- [x] [Server] Attach the quiz questions to the quiz exercises

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration -> Exam
3. Create exercises grouped liked seen in the screenshots
--> Verify that the information is displayed correctly
4. Delete exercises in the mixed group one-by-one
--> Verify that the table only shows columns relevant for the existing exercises

### Screenshots
group_mixed
Mixed group

group_mod
Only modeling

group_programming
Only programming

group_quiz
Only quiz

group_text
Only text

### CODE REVIEW

**Comment by krusche** [comment]
The screenshot shows an exercise group with mixed exercise types.
Typically all exercises in a group have the same type. In such a case some columns are not necessary.
Can we show the columns for specific types only if there is at least one exercise of this type in the exercise group?

I think this would increase the readability of the screen.

Also please add additional information which would be interesting.
* Quiz: Number of questions --> we should add this in a course as well
* Text: Assessment Mode (Manual vs. Semi-automatic) --> we should add this information in the course exercises overview as well
* Modeling: Assessment Mode (Manuel vs. Semi-automatic) --> we should add this information in the course exercises overview as well
* Programming: Short name, Assessment Mode (Automatic vs. Semi-automatic), Hidden tests (this is either false in case build and run test after due date is not set, or it shows the current number of hidden tests compared to all tests, e.g. 5/10 and the date when all tests will be executed automatically)  --> we should add this information in the course exercises overview as well

**Comment by krusche** [comment]
I guess some of the information of my last comment is a bit more complex to calculate, e.g. the one with the hidden test.
I think we should go for the easy information to get displayed. We can display the more complex one in a follow up

**Comment by kloessst** [comment]
I've extended the exercise group view with the requested columns.
Extension of the course view will be done in a follow up.

File: src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.ts [hunk 1]
```diff
@@ -84,6 +87,34 @@ export class ExerciseGroupsComponent implements OnInit {
...
+        return null;
+    }
+
+    asModelingExercise(exercise: Exercise): ModelingExercise | null {
+        if (exercise.type === ExerciseType.MODELING) {
+            return exercise as ModelingExercise;
+        }
+        return null;
+    }
+
```
Comment (sascha11110) [inline src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.ts:117]: I think this is optimal for a switch case. As we have an enum `ExerciseType` defined we could use that to be type-safe for the second parameter of the function:

```suggestion
    caseExercise(exercise: Exercise, exerciseType: ExerciseType): ProgrammingExercise | QuizExercise | TextExercise | ModelingExercise | null {
        switch (exerciseType) {
            case ExerciseType.PROGRAMMING:
                 return exercise as ProgrammingExercise;
            ...
        }
        return null;
    }
    
```

**Review by sascha11110 — APPROVED** [review]
I've test it on the test server and it works quite well! I think you have addressed all easy to implement suggestions of @krusche's comment.

Code-wise it looks good to me also, I've just added a smaller suggestion.

File: src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.html [hunk 2]
```diff
@@ -154,6 +186,143 @@ <h5 class="font-weight-bold mb-0">{{ exerciseGroup.title }}</h5>
...
+                                                target="_blank"
+                                                >Template</a
+                                            >
+                                        </span>
+                                        <ng-template #noVersionControlAndContinuousIntegrationAvailableTemplate>Template</ng-template>
+                                        <jhi-programming-exercise-instructor-status
+                                            *ngIf="
+                                                programmingExercise.templateParticipation &&
+                                                programmingExercise.templateParticipation.results &&
+                                                programmingExercise.templateParticipation.results.length
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.html:215]:
```suggestion
  
                                                programmingExercise?.templateParticipation?.results?.length
```
I think optional chaining should work here

File: src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.html [hunk 3]
```diff
@@ -154,6 +186,143 @@ <h5 class="font-weight-bold mb-0">{{ exerciseGroup.title }}</h5>
...
+                                                target="_blank"
+                                                >Solution</a
+                                            >
+                                        </span>
+                                        <ng-template #noVersionControlAndContinuousIntegrationAvailableSolution>Solution</ng-template>
+                                        <jhi-programming-exercise-instructor-status
+                                            *ngIf="
+                                                programmingExercise.solutionParticipation &&
+                                                programmingExercise.solutionParticipation.results &&
+                                                programmingExercise.solutionParticipation.results.length
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.html:239]:
```suggestion
programmingExercise?. solutionParticipation?.results?.length
```

File: src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.ts (hunk 1 above)
Comment (kloessst) [reply src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.ts:117]: The casting methods above are used like:
`<ng-container *ngIf="asSpecificExercise(exercise) as programmingExercise">`
They fulfill two purposes:
1. Cast to a programming exercise if it is one
2. Return null if it isn't a programming exercise

If we have a modeling exercise for example, the switch case statement would do a cast to the modeling exercise but won't return null. ngIf will be true.
Maybe we don't need the casting at all. We won't have typescript tooling support in the template then.

Comment (kloessst) [reply src/main/webapp/app/exam/manage/exercise-groups/exercise-groups.component.ts:117]: https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions
We can stick to just checking just the exercise.type in the template

**Review by sascha11110 — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Comment by krusche** [comment]
I'm not 100% happy with the code quality, but I think we can improve this in a follow-up in the future, when we display more information. I would suggest we try to extract Angular components for each exercise type to sync how the information is displayed for exam exercises and for course exercises.
Would be nice if this can be reused in the future so that changes to one of them are also reflected for the other one!

**Comment by krusche** [comment]
@anditurdiu and @kloessst please put this on your TODO list!

**Pull request merged**